gateway-ha/src/main/java/com/lyft/data/gateway/ha/handler/QueryIdCachingProxyHandler.java

```diff
@@ -115,22 +116,23 @@ public String rewriteTarget(HttpServletRequest request) {
 
     // Only load balance presto query APIs.
     if (isPathWhiteListed(request.getRequestURI())) {
-      String queryId = extractQueryIdIfPresent(request);
+      Optional<String> queryId = extractQueryIdIfPresent(request);
 
+      backendAddress = queryId
       // Find query id and get url from cache
-      if (!Strings.isNullOrEmpty(queryId)) {
-        backendAddress = routingManager.findBackendForQueryId(queryId);
-      } else {
+      .map(routingManager::findBackendForQueryId)
+      .orElseGet(() -> {
         String routingGroup = routingGroupSelector.findRoutingGroup(request);
         String user = Optional.ofNullable(request.getHeader(USER_HEADER))
                 .orElse(request.getHeader(ALTERNATE_USER_HEADER));
         if (!Strings.isNullOrEmpty(routingGroup)) {
           // This falls back on adhoc backend if there are no cluster found for the routing group.
-          backendAddress = routingManager.provideBackendForRoutingGroup(routingGroup, user);
+          return routingManager.provideBackendForRoutingGroup(routingGroup, user);
```

Comment:
please dont return from here.  Below lines are skipped. without this, subsequent calls would require a broadcast to all backends for query id lookup.
```
       // set target backend so that we could save queryId to backend mapping later.
      ((MultiReadHttpServletRequest) request).addHeader(PROXY_TARGET_HEADER, backendAddress);
```